From c509a37deef9335570c570903190f5821654bec9 Mon Sep 17 00:00:00 2001
From: Bruno Martins <bgcngm@gmail.com>
Date: Mon, 25 Dec 2017 21:16:28 +0000
Subject: [PATCH 3/8] lineage-sdk: Import device keys custom rebinding configs
 and add helpers

 * Used to "live" in fw/b since ever, but will be here from now on.

 * This also brings AOSP configs (config_longPressOnHomeBehavior and
   config_doubleTapOnHomeBehavior) to the SDK since those are being
   extended, putting everything in one place.

Change-Id: I0ca2b2cf9a565a334c3edf19a978fa6d2c4d2ea3
---
 lineage/res/res/values-large/config.xml            | 21 +++++++
 lineage/res/res/values-sw600dp/config.xml          | 21 +++++++
 lineage/res/res/values/config.xml                  | 64 ++++++++++++++++++++++
 lineage/res/res/values/symbols.xml                 |  6 ++
 .../internal/util/DeviceKeysConstants.java         | 35 ++++++++++++
 5 files changed, 147 insertions(+)
 create mode 100644 lineage/res/res/values-large/config.xml
 create mode 100644 lineage/res/res/values-sw600dp/config.xml

diff --git a/lineage/res/res/values-large/config.xml b/lineage/res/res/values-large/config.xml
new file mode 100644
index 0000000..bc09b4a
--- /dev/null
+++ b/lineage/res/res/values-large/config.xml
@@ -0,0 +1,21 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!--
+     Copyright (C) 2015 The CyanogenMod Project
+                   2017 The LineageOS Project
+
+     Licensed under the Apache License, Version 2.0 (the "License");
+     you may not use this file except in compliance with the License.
+     You may obtain a copy of the License at
+
+          http://www.apache.org/licenses/LICENSE-2.0
+
+     Unless required by applicable law or agreed to in writing, software
+     distributed under the License is distributed on an "AS IS" BASIS,
+     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+     See the License for the specific language governing permissions and
+     limitations under the License.
+-->
+<resources>
+    <!-- see comment in values/config.xml -->
+    <integer name="config_longPressOnHomeBehavior">0</integer>
+</resources>
diff --git a/lineage/res/res/values-sw600dp/config.xml b/lineage/res/res/values-sw600dp/config.xml
new file mode 100644
index 0000000..bc09b4a
--- /dev/null
+++ b/lineage/res/res/values-sw600dp/config.xml
@@ -0,0 +1,21 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!--
+     Copyright (C) 2015 The CyanogenMod Project
+                   2017 The LineageOS Project
+
+     Licensed under the Apache License, Version 2.0 (the "License");
+     you may not use this file except in compliance with the License.
+     You may obtain a copy of the License at
+
+          http://www.apache.org/licenses/LICENSE-2.0
+
+     Unless required by applicable law or agreed to in writing, software
+     distributed under the License is distributed on an "AS IS" BASIS,
+     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+     See the License for the specific language governing permissions and
+     limitations under the License.
+-->
+<resources>
+    <!-- see comment in values/config.xml -->
+    <integer name="config_longPressOnHomeBehavior">0</integer>
+</resources>
diff --git a/lineage/res/res/values/config.xml b/lineage/res/res/values/config.xml
index cdd8d4e..b12a8c7 100644
--- a/lineage/res/res/values/config.xml
+++ b/lineage/res/res/values/config.xml
@@ -168,4 +168,68 @@
          For example, a device with Home, Back and Menu keys would set this
          config to 7. -->
     <integer name="config_deviceHardwareWakeKeys">79</integer>
+
+    <!-- Control the behavior when the user long presses the home button.
+            0 - Nothing
+            1 - Menu key
+            2 - Recent apps view in SystemUI
+            3 - Launch assist intent
+            4 - Voice Search
+            5 - In-app Search
+            6 - Launch camera
+            7 - Sleep
+            8 - Last app
+            9 - Toggle split screen
+         This needs to match the enums in
+         sdk/src/java/org/lineageos/internal/util/DeviceKeysConstants.java.
+    -->
+    <integer name="config_longPressOnHomeBehavior">2</integer>
+
+    <!-- Control the behavior when the user double-taps the home button.
+            0 - Nothing
+            1 - Menu key
+            2 - Recent apps view in SystemUI
+            3 - Launch assist intent
+            4 - Voice Search
+            5 - In-app Search
+            6 - Launch camera
+            7 - Sleep
+            8 - Last app
+            9 - Toggle split screen
+         This needs to match the enums in
+         sdk/src/java/org/lineageos/internal/util/DeviceKeysConstants.java.
+    -->
+    <integer name="config_doubleTapOnHomeBehavior">0</integer>
+
+    <!-- Control the behavior when the user long presses the menu button.
+            0 - Nothing
+            1 - Menu key
+            2 - Recent apps view in SystemUI
+            3 - Launch assist intent
+            4 - Voice Search
+            5 - In-app Search
+            6 - Launch camera
+            7 - Sleep
+            8 - Last app
+            9 - Toggle split screen
+         This needs to match the enums in
+         sdk/src/java/org/lineageos/internal/util/DeviceKeysConstants.java.
+    -->
+    <integer name="config_longPressOnMenuBehavior">3</integer>
+
+    <!-- Control the behavior when the user long presses the app switch button.
+            0 - Nothing
+            1 - Menu key
+            2 - Recent apps view in SystemUI
+            3 - Launch assist intent
+            4 - Voice Search
+            5 - In-app Search
+            6 - Launch camera
+            7 - Sleep
+            8 - Last app
+            9 - Toggle split screen
+         This needs to match the enums in
+         sdk/src/java/org/lineageos/internal/util/DeviceKeysConstants.java.
+    -->
+    <integer name="config_longPressOnAppSwitchBehavior">9</integer>
 </resources>
diff --git a/lineage/res/res/values/symbols.xml b/lineage/res/res/values/symbols.xml
index 6ffd51f..fbce76d 100644
--- a/lineage/res/res/values/symbols.xml
+++ b/lineage/res/res/values/symbols.xml
@@ -120,6 +120,12 @@
     <java-symbol type="integer" name="config_deviceHardwareKeys" />
     <java-symbol type="integer" name="config_deviceHardwareWakeKeys" />
 
+    <!-- Device keys user-customisable behavior -->
+    <java-symbol type="integer" name="config_longPressOnHomeBehavior" />
+    <java-symbol type="integer" name="config_doubleTapOnHomeBehavior" />
+    <java-symbol type="integer" name="config_longPressOnMenuBehavior" />
+    <java-symbol type="integer" name="config_longPressOnAppSwitchBehavior" />
+
     <!-- BurnIn protection -->
     <java-symbol type="bool" name="config_enableBurnInProtection" />
 
diff --git a/sdk/src/java/org/lineageos/internal/util/DeviceKeysConstants.java b/sdk/src/java/org/lineageos/internal/util/DeviceKeysConstants.java
index 053afa9..17d7e70 100644
--- a/sdk/src/java/org/lineageos/internal/util/DeviceKeysConstants.java
+++ b/sdk/src/java/org/lineageos/internal/util/DeviceKeysConstants.java
@@ -16,7 +16,42 @@
 
 package org.lineageos.internal.util;
 
+import android.content.ContentResolver;
+import android.os.UserHandle;
+
+import lineageos.providers.LineageSettings;
+
 public class DeviceKeysConstants {
+    // Available custom actions to perform on a key press.
+    // Must match values for KEY_HOME_LONG_PRESS_ACTION in:
+    //   sdk/src/java/lineageos/providers/LineageSettings.java
+    public enum Action {
+        NOTHING,
+        MENU,
+        APP_SWITCH,
+        SEARCH,
+        VOICE_SEARCH,
+        IN_APP_SEARCH,
+        LAUNCH_CAMERA,
+        SLEEP,
+        LAST_APP,
+        SPLIT_SCREEN,
+        SINGLE_HAND_LEFT,
+        SINGLE_HAND_RIGHT;
+
+        public static Action fromIntSafe(int id) {
+            if (id < NOTHING.ordinal() || id > Action.values().length) {
+                return NOTHING;
+            }
+            return Action.values()[id];
+        }
+
+        public static Action fromSettings(ContentResolver cr, String setting, Action def) {
+            return fromIntSafe(LineageSettings.System.getIntForUser(cr,
+                    setting, def.ordinal(), UserHandle.USER_CURRENT));
+        }
+    }
+
     // Masks for checking presence of hardware keys.
     // Must match values in:
     //   lineage/res/res/values/config.xml
-- 
2.7.4

